synapse/storage/databases/state/store.py

```diff
@@ -503,13 +525,32 @@ def _store_state_group_txn(txn: LoggingTransaction) -> int:
                 self._state_group_cache.update,
                 self._state_group_cache.sequence,
                 key=state_group,
-                value=dict(current_non_member_state_ids),
+                value=current_non_member_state_ids,
             )
 
             return state_group
 
+        if prev_group is not None:
+            state_group = await self.db_pool.runInteraction(
+                "store_state_group.insert_delta_group",
+                insert_delta_group_txn,
+                prev_group,
+                delta_ids,
+            )
+            if state_group is not None:
+                return state_group
+
+        if current_state_ids is None:
```

Comment:
```suggestion
       # We're going to persist the state as a complete group rather than
       # a delta, so first we need to ensure we have loaded the state map
       # from the database.
        if current_state_ids is None:
```